core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -0,0 +1,343 @@
+
+        if (visited.containsKey(valueType.signature())) {
+            final String pathName = visited.get(valueType.signature());
+            additionalProperties.put("$ref", pathName);
+        } else {
+            additionalProperties.put("type", innerSchemaType);
+
+            if ("object".equals(innerSchemaType)) {
+                final StructInfo fieldStructInfo = typeNameToStructMapping.get(valueType.name());
+
+                visited.put(valueType.signature(), nextPath);
+                generateFields(fieldStructInfo.fields(), visited, nextPath + "/additionalProperties",
+                               additionalProperties);
+            }
+        }
+
+        fieldNode.set("additionalProperties", additionalProperties);
+    }
+
+    private void generateArrayFields(ObjectNode fieldNode, FieldInfo field, Map<String, String> visited,
+                                     String path) {
+        final ObjectNode items = mapper.createObjectNode();
+        final String nextPath = path + '/' + field.name();
+
+        final TypeSignature itemsType =
+                ((ContainerTypeSignature) field.typeSignature()).typeParameters().get(0);
+
+        final String innerSchemaType = getSchemaType(itemsType);
+
+        if (visited.containsKey(itemsType.signature())) {
+            final String pathName = visited.get(itemsType.signature());
+            items.put("$ref", pathName);
+        } else {
+            items.put("type", innerSchemaType);
+
+            if ("object".equals(innerSchemaType)) {
+                final StructInfo fieldStructInfo = typeNameToStructMapping.get(itemsType.name());
+                visited.put(itemsType.signature(), nextPath);
+                generateFields(fieldStructInfo.fields(), visited, nextPath + "/items", items);
+            }
+        }
+
+        fieldNode.set("items", items);
+    }
+
+    private void generateStructFields(ObjectNode fieldNode, FieldInfo field, Map<String, String> visited,
+                                      String path) {
+        fieldNode.put("additionalProperties", true);
+
+        final StructInfo fieldStructInfo = typeNameToStructMapping.get(field.typeSignature().name());
+
+        // Iterate over each child field, generate their definitions.
+        if (fieldStructInfo != null && !fieldStructInfo.fields().isEmpty()) {
+            generateFields(fieldStructInfo.fields(), visited, path, fieldNode);
+        }
+    }
+
+    @Nullable
+    private ArrayNode getEnumType(TypeSignature type) {
+        final ArrayNode enumArray = mapper.createArrayNode();
+        final EnumInfo enumInfo = typeNameToEnumMapping.get(type.signature());
+
+        if (enumInfo != null) {
+            enumInfo.values().forEach(x -> enumArray.add(x.name()));
+        }
+
+        return enumArray;
+    }
+
+    private String getSchemaType(TypeSignature typeSignature) {
+        if (typeSignature.type() == TypeSignatureType.ENUM) {
+            return "string";
```

Comment:
Is this string?

Reply:
Yes, enums are `string` with an additional field called `enum` that contains possible string values it takes.